golangci-lint doesn't open quickfix or location lists?

### What did you do? (required: The issue will be **closed** when not provided)

Opened a `.go` file, made some code changes and then `:w` to write the buffer back to the file. This caused the message `vim-go: [golangci-lint] FAIL` to be displayed, but the information about what is failing isn't being shown. 

### What did you expect to happen?

I expected the quickfix window to open and to have a list of errors contained within it.

### What happened instead?

The status bar indicated there was an error `vim-go: [golangci-lint] FAIL` but it didn't open the quickfix list (nor the location list).

> Note: I manually opened both `:copen` and `:lopen` just in case the errors were listed there but for some reason the vim-go plugin had failed to simply 'open' the relevant list window.

### Configuration (**MUST** fill this out):

```viml
Plug 'fatih/vim-go', { 'do': ':GoUpdateBinaries' }

...

let g:go_fmt_command = 'goimports'
let g:go_metalinter_autosave = 1
let g:go_metalinter_deadline = '20s'

autocmd FileType go imap /. <C-x><C-o>
```

#### vim-go version:

v1.22

####  `vimrc` you used to reproduce (use a *minimal* vimrc with other plugins disabled; do not link to a 2,000 line vimrc):
<details><summary>vimrc</summary><br><pre>
set nocompatible
set nobackup
set nowritebackup
set noswapfile

call plug#begin('~/.vim/plugged')

Plug 'fatih/vim-go', { 'do': ':GoUpdateBinaries' }

call plug#end()

let g:go_fmt_command = 'goimports'
let g:go_metalinter_autosave = 1
let g:go_metalinter_deadline = '20s'

autocmd FileType go imap /. <C-x><C-o>
</pre></details>

#### Vim version (first three lines from `:version`):

```
VIM - Vi IMproved 8.1 (2018 May 18, compiled Aug 12 2019 11:40:34)
macOS version
Included patches: 1-1839
```

#### Go version (`go version`):

```
go version go1.13 darwin/amd64
```

#### Go environment
<details><summary><code>go env</code> Output:</summary><br><pre>
<!-- go env -->
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/markmcdonnell/Library/Caches/go-build"
GOENV="/Users/markmcdonnell/Library/Application Support/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/markmcdonnell/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/Cellar/go/1.13/libexec"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/Cellar/go/1.13/libexec/pkg/tool/darwin_amd64"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/zk/_jgd9h392p55w9fd4lzt5dpw0000gp/T/go-build183558573=/tmp/go-build -gno-record-gcc-switches -fno-common"
</pre></details>

#### gopls version
<details><summary><code>gopls version</code> Output:</summary><br><pre>
<!-- gopls version -->
golang.org/x/tools/gopls v0.3.1
    golang.org/x/tools/gopls@v0.3.1 h1:yNTWrf4gc4Or0UecjOas5pzOa3BL0WDDyKDV4Wz5VaM=
</pre></details>

> What do you mean by "when the code is bad"? Do you mean it won't compile?

So when I say "the code is bad" I mean the code either wouldn't compile or in some rare occasions the code would have violated a 'style' type of lint check (but mainly it's code that wont  compile). 

> What happens if you run the golangci-lint command that vim-go uses 

I've tested this with a reduced test case file that I would expect to raise a compiler error...

```
$ /Users/markmcdonnell/go/bin/golangci-lint run --print-issued-lines=false --build-tags --disable-all --exclude-use-default=false --enable=govet --enable=golint --deadline=20s /Users/markmcdonnell/Code/Go

WARN [runner] Can't run linter unused: buildssa: analysis skipped: errors in package: [-: import "/Users/markmcdonnell/Code/Go": cannot import absolute path]
ERRO [runner/goanalysis_metalinter] typechecking error: import "/Users/markmcdonnell/Code/Go": cannot import absolute path
WARN [runner] Can't run linter goanalysis_metalinter: assign: failed prerequisites: inspect@/Users/markmcdonnell/Code/Go
ERRO Running error: assign: failed prerequisites: inspect@/Users/markmcdonnell/Code/Go
```

I've also tested this with some 'real' code, and inside vim when I `:w` the file the response from the golangci-lint is a `SUCCESS'. But running the cli command suggests a failure...

```
/Users/markmcdonnell/go/bin/golangci-lint run --print-issued-lines=false --build-tags --disable-all --exclude-use-default=false --enable=govet --enable=golint --deadline=20s /Users/markmcdonnell/Code/buzzfeed/mono/perimeter/internal/pkg/proxy

ERRO [runner/goanalysis_metalinter] typechecking error: import "/Users/markmcdonnell/Code/buzzfeed/mono/perimeter/internal/pkg/proxy": cannot import absolute path
WARN [runner] Can't run linter goanalysis_metalinter: fact_purity: failed prerequisites: buildssa@/Users/markmcdonnell/Code/buzzfeed/mono/perimeter/internal/pkg/proxy
WARN [runner] Can't run linter unused: buildssa: analysis skipped: errors in package: [-: import "/Users/markmcdonnell/Code/buzzfeed/mono/perimeter/internal/pkg/proxy": cannot import absolute path]
ERRO Running error: buildssa: analysis skipped: errors in package: [-: import "/Users/markmcdonnell/Code/buzzfeed/mono/perimeter/internal/pkg/proxy": cannot import absolute path]
```

I then remembered that in the 'real' example I'm working from a mono repo and so I open vim from the root directory and then `:lcd ./path/to/service` so I decided to `cd` directly to the directory (in case `:lcd` was confusing the linter somehow) and found a different response from running the cli command...

```
$ /Users/markmcdonnell/go/bin/golangci-lint run --print-issued-lines=false --build-tags --disable-all --exclude-use-default=false --enable=govet --enable=golint --deadline=20s /Users/markmcdonnell/Code/buzzfeed/mono/perimeter/internal/pkg/proxy

internal/pkg/proxy/proxy_test.go:168:25: Error return value of `resp.Body.Close` is not checked (errcheck)
internal/pkg/proxy/proxy_test.go:254:25: Error return value of `resp.Body.Close` is not checked (errcheck)
internal/pkg/proxy/proxy_test.go:319:23: Error return value of `resp.Body.Close` is not checked (errcheck)
internal/pkg/proxy/static.go:52:2: S1005: should write `sp := request.StaticPaths[r.URL.Path]` instead of `sp, _ := request.StaticPaths[r.URL.Path]` (gosimple)
internal/pkg/proxy/base.go:122:2: S1008: should use 'return <expr>' instead of 'if <expr> { return <bool> }; return <bool>' (gosimple)
internal/pkg/proxy/base.go:132:2: S1008: should use 'return <expr>' instead of 'if <expr> { return <bool> }; return <bool>' (gosimple)
internal/pkg/proxy/base.go:358:2: S1008: should use 'return <expr>' instead of 'if <expr> { return <bool> }; return <bool>' (gosimple)
internal/pkg/proxy/base.go:648:5: S1002: should omit comparison to bool constant, can be simplified to `rh.storage.Stopped` (gosimple)
```

I then decided to open up vim from the actual directory where the code was and to break the code such that it wouldn't compile. But again, the FAIL message was displayed but no list was opened with the error and the `:messages` didn't show it either.

Thanks @Integralist . That's all really helpful. I can explain why the these kinds of lines aren't showing up:
```
ERRO [runner/goanalysis_metalinter] typechecking error: import "/Users/markmcdonnell/Code/buzzfeed/mono/perimeter/internal/pkg/proxy": cannot import absolute path
WARN [runner] Can't run linter goanalysis_metalinter: fact_purity: failed prerequisites: buildssa@/Users/markmcdonnell/Code/buzzfeed/mono/perimeter/internal/pkg/proxy
WARN [runner] Can't run linter unused: buildssa: analysis skipped: errors in package: [-: import "/Users/markmcdonnell/Code/buzzfeed/mono/perimeter/internal/pkg/proxy": cannot import absolute path]
ERRO Running error: buildssa: analysis skipped: errors in package: [-: import "/Users/markmcdonnell/Code/buzzfeed/mono/perimeter/internal/pkg/proxy": cannot import absolute path]
```

Those lines won't use the [error format string](https://github.com/fatih/vim-go/blob/master/autoload/go/lint.vim#L346) used by vim-go. I'm working on a patch now to fix that.

These kinds of lines _should_ be showing up, though:
```
internal/pkg/proxy/proxy_test.go:168:25: Error return value of `resp.Body.Close` is not checked (errcheck)
internal/pkg/proxy/proxy_test.go:254:25: Error return value of `resp.Body.Close` is not checked (errcheck)
internal/pkg/proxy/proxy_test.go:319:23: Error return value of `resp.Body.Close` is not checked (errcheck)
internal/pkg/proxy/static.go:52:2: S1005: should write `sp := request.StaticPaths[r.URL.Path]` instead of `sp, _ := request.StaticPaths[r.URL.Path]` (gosimple)
internal/pkg/proxy/base.go:122:2: S1008: should use 'return <expr>' instead of 'if <expr> { return <bool> }; return <bool>' (gosimple)
internal/pkg/proxy/base.go:132:2: S1008: should use 'return <expr>' instead of 'if <expr> { return <bool> }; return <bool>' (gosimple)
internal/pkg/proxy/base.go:358:2: S1008: should use 'return <expr>' instead of 'if <expr> { return <bool> }; return <bool>' (gosimple)
internal/pkg/proxy/base.go:648:5: S1002: should omit comparison to bool constant, can be simplified to `rh.storage.Stopped` (gosimple)
```

🤔

@bhcleek I'm not sure it is related to this PR, but I'll ask here before submitting a new issue, because it is also related to golangci-lint output. In my case, quickfix window is populated with the results, but I can't go to any location from the list.

For example, the output is always like this:
```
|| util.go:8:60: `buildReply` - `userToken` always receives `"N/A"` (unparam)
|| func buildReply(status bool, command string, reply string, userToken string) []byte {
```
(note the `||` at the beginning). I used to be able to navigate to each warning from the list some time ago, and then it stopped working. Is it related to this issue? Should I open a new one with more details?

--- a/autoload/go/lint.vim
+++ b/autoload/go/lint.vim
@@ -331,7 +331,10 @@ function! s:metalinterautosavecomplete(filepath, job, exit_code, messages)
 
   let l:idx = len(a:messages) - 1
   while l:idx >= 0
-    if a:messages[l:idx] !~# '^' . a:filepath . ':'
+    " leave in any messages that report errors about a:filepath or that report
+    " more general problems that prevent golangci-lint from linting
+    " a:filepath.
+    if a:messages[l:idx] !~# '^' . a:filepath . ':' && a:messages[l:idx] !~# '^level='
       call remove(a:messages, l:idx)
     endif
     let l:idx -= 1
@@ -343,7 +346,7 @@ function! s:errorformat(metalinter) abort
     " Golangci-lint can output the following:
     "   <file>:<line>:<column>: <message> (<linter>)
     " This can be defined by the following errorformat:
-    return '%f:%l:%c:\ %m'
+    return 'level=%tarning\ msg="%m:\ [%f:%l:%c:\ %.%#]",level=%trror\ msg="%m:\ [%f:%l:%c:\ %.%#]",%f:%l:%c:\ %m'
   elseif a:metalinter == 'gopls'
     return '%f:%l:%c:%t:\ %m,%f:%l:%c::\ %m'
   endif

--- a/autoload/go/lint.vim
+++ b/autoload/go/lint.vim
@@ -68,7 +68,7 @@ function! go#lint#Gometa(bang, autosave, ...) abort
     let l:err = len(l:messages)
   else
     if go#util#has_job()
-      call s:lint_job({'cmd': cmd, 'statustype': l:metalinter, 'errformat': errformat}, a:bang, a:autosave)
+      call s:lint_job(l:metalinter, {'cmd': cmd, 'statustype': l:metalinter, 'errformat': errformat}, a:bang, a:autosave)
       return
     endif
 
@@ -89,8 +89,8 @@ function! go#lint#Gometa(bang, autosave, ...) abort
     let l:winid = win_getid(winnr())
     " Parse and populate our location list
 
-    if a:autosave && l:metalinter != 'gopls'
-      call s:metalinterautosavecomplete(fnamemodify(expand('%:p'), ":."), 0, 1, l:messages)
+    if a:autosave
+      call s:metalinterautosavecomplete(l:metalinter, fnamemodify(expand('%:p'), ":."), 0, 1, l:messages)
     endif
     call go#list#ParseFormat(l:listtype, errformat, l:messages, 'GoMetaLinter')
 
@@ -278,7 +278,7 @@ function! go#lint#ToggleMetaLinterAutoSave() abort
   call go#util#EchoProgress("auto metalinter enabled")
 endfunction
 
-function! s:lint_job(args, bang, autosave)
+function! s:lint_job(metalinter, args, bang, autosave)
   let l:opts = {
         \ 'statustype': a:args.statustype,
         \ 'errorformat': a:args.errformat,
@@ -289,7 +289,7 @@ function! s:lint_job(args, bang, autosave)
   if a:autosave
     let l:opts.for = "GoMetaLinterAutoSave"
     " s:metalinterautosavecomplete is really only needed for golangci-lint
-    let l:opts.complete = funcref('s:metalinterautosavecomplete', [expand('%:p:t')])
+    let l:opts.complete = funcref('s:metalinterautosavecomplete', [a:metalinter, expand('%:p:t')])
   endif
 
   " autowrite is not enabled for jobs
@@ -324,7 +324,11 @@ function! s:golangcilintcmd(bin_path)
   return cmd
 endfunction
 
-function! s:metalinterautosavecomplete(filepath, job, exit_code, messages)
+function! s:metalinterautosavecomplete(metalinter, filepath, job, exit_code, messages)
+  if a:metalinter != 'golangci-lint'
+    return
+  endif
+
   if len(a:messages) == 0
     return
   endif

--- a/autoload/go/lint_test.vim
+++ b/autoload/go/lint_test.vim
@@ -17,7 +17,7 @@ func! s:gometa(metalinter) abort
         \ ]
     if a:metalinter == 'golangci-lint'
       let expected = [
-            \ {'lnum': 5, 'bufnr': bufnr('%')+2, 'col': 1, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'exported function `MissingFooDoc` should have comment or be unexported (golint)'}
+            \ {'lnum': 5, 'bufnr': bufnr('%')+4, 'col': 1, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'exported function `MissingFooDoc` should have comment or be unexported (golint)'}
           \ ]
     endif
 
@@ -27,6 +27,42 @@ func! s:gometa(metalinter) abort
     let g:go_metalinter_enabled = ['golint']
 
     call go#lint#Gometa(0, 0, $GOPATH . '/src/foo')
+
+    let actual = getqflist()
+    let start = reltime()
+    while len(actual) == 0 && reltimefloat(reltime(start)) < 10
+      sleep 100m
+      let actual = getqflist()
+    endwhile
+
+    call gotest#assert_quickfix(actual, expected)
+  finally
+      call call(RestoreGOPATH, [])
+      unlet g:go_metalinter_enabled
+  endtry
+endfunc
+
+func! Test_GometaGolangciLint_problems() abort
+  call s:gometa_problems('golangci-lint')
+endfunc
+
+func! s:gometa_problems(metalinter) abort
+  let RestoreGOPATH = go#util#SetEnv('GOPATH', fnamemodify(getcwd(), ':p') . 'test-fixtures/lint')
+  silent exe 'e ' . $GOPATH . '/src/lint/golangci-lint/problems/problems.go'
+
+  try
+    let g:go_metalinter_command = a:metalinter
+    let expected = [
+          \ {'lnum': 3, 'bufnr': bufnr('%'), 'col': 8, 'pattern': '', 'valid': 1, 'vcol': 0, 'nr': -1, 'type': 'w', 'module': '', 'text': '[runner] Can''t run linter golint: golint: analysis skipped: errors in package'},
+          \ {'lnum': 3, 'bufnr': bufnr('%'), 'col': 8, 'pattern': '', 'valid': 1, 'vcol': 0, 'nr': -1, 'type': 'e', 'module': '', 'text': 'Running error: golint: analysis skipped: errors in package'}
+        \ ]
+
+    " clear the quickfix lists
+    call setqflist([], 'r')
+
+    let g:go_metalinter_enabled = ['golint']
+
+    call go#lint#Gometa(0, 0)
 
     let actual = getqflist()
     let start = reltime()
@@ -60,6 +96,42 @@ func! s:gometaautosave(metalinter) abort
             \ {'lnum': 5, 'bufnr': bufnr('%'), 'col': 1, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'exported function `MissingDoc` should have comment or be unexported (golint)'}
           \ ]
     endif
+
+    " clear the location lists
+    call setloclist(0, [], 'r')
+
+    let g:go_metalinter_autosave_enabled = ['golint']
+
+    call go#lint#Gometa(0, 1)
+
+    let actual = getloclist(0)
+    let start = reltime()
+    while len(actual) == 0 && reltimefloat(reltime(start)) < 10
+      sleep 100m
+      let actual = getloclist(0)
+    endwhile
+
+    call gotest#assert_quickfix(actual, expected)
+  finally
+    call call(RestoreGOPATH, [])
+    unlet g:go_metalinter_autosave_enabled
+  endtry
+endfunc
+
+func! Test_GometaAutoSaveGolangciLint_problems() abort
+  call s:gometaautosave_problems('golangci-lint')
+endfunc
+
+func! s:gometaautosave_problems(metalinter) abort
+  let RestoreGOPATH = go#util#SetEnv('GOPATH', fnameescape(fnamemodify(getcwd(), ':p')) . 'test-fixtures/lint')
+  silent exe 'e ' . $GOPATH . '/src/lint/golangci-lint/problems/ok.go'
+
+  try
+    let g:go_metalinter_command = a:metalinter
+    let expected = [
+          \ {'lnum': 3, 'bufnr': bufnr('%')+1, 'col': 8, 'pattern': '', 'valid': 1, 'vcol': 0, 'nr': -1, 'type': 'w', 'module': '', 'text': '[runner] Can''t run linter golint: golint: analysis skipped: errors in package'},
+          \ {'lnum': 3, 'bufnr': bufnr('%')+1, 'col': 8, 'pattern': '', 'valid': 1, 'vcol': 0, 'nr': -1, 'type': 'e', 'module': '', 'text': 'Running error: golint: analysis skipped: errors in package'}
+        \ ]
 
     " clear the location lists
     call setloclist(0, [], 'r')
@@ -149,7 +221,7 @@ func! Test_Lint_GOPATH() abort
 
   let expected = [
           \ {'lnum': 5, 'bufnr': bufnr('%'), 'col': 1, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'exported function MissingDoc should have comment or be unexported'},
-          \ {'lnum': 5, 'bufnr': bufnr('%')+4, 'col': 1, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'exported function AlsoMissingDoc should have comment or be unexported'}
+          \ {'lnum': 5, 'bufnr': bufnr('%')+6, 'col': 1, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'exported function AlsoMissingDoc should have comment or be unexported'}
       \ ]
 
   let winnr = winnr()
@@ -177,7 +249,7 @@ func! Test_Lint_NullModule() abort
 
   let expected = [
           \ {'lnum': 5, 'bufnr': bufnr('%'), 'col': 1, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'exported function MissingDoc should have comment or be unexported'},
-          \ {'lnum': 5, 'bufnr': bufnr('%')+4, 'col': 1, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'exported function AlsoMissingDoc should have comment or be unexported'}
+          \ {'lnum': 5, 'bufnr': bufnr('%')+6, 'col': 1, 'valid': 1, 'vcol': 0, 'nr': -1, 'type': '', 'pattern': '', 'text': 'exported function AlsoMissingDoc should have comment or be unexported'}
       \ ]
 
   let winnr = winnr()

--- a/autoload/go/test-fixtures/lint/src/lint/baz.go
+++ b/autoload/go/test-fixtures/lint/src/lint/baz.go
@@ -1,0 +1,3 @@
+package lint
+
+func baz() {}

--- a/autoload/go/test-fixtures/lint/src/lint/golangci-lint/problems/ok.go
+++ b/autoload/go/test-fixtures/lint/src/lint/golangci-lint/problems/ok.go
@@ -1,0 +1,3 @@
+package problems
+
+func bar() {}

--- a/autoload/go/test-fixtures/lint/src/lint/golangci-lint/problems/problems.go
+++ b/autoload/go/test-fixtures/lint/src/lint/golangci-lint/problems/problems.go
@@ -1,0 +1,5 @@
+package problems
+
+import "/quux"
+
+func baz() {}

